priorityq.py

```diff
@@ -4,57 +4,103 @@
 
 class Element(object):
     ''' Create data element with default value and next pointer. '''
-    def __init__(self, value, priority=0, behind=None):
+    def __init__(self, value, priority=0):
         ''' Value and next pointer default to none. '''
         self.val = value
-        self.behind = behind
         self.prio = priority
 
 
 class Priorityq(object):
-    ''' Create an empty queue. '''
+    ''' Create an empty heap. '''
     def __init__(self):
-        self.front = None
-        self.back = None
+        self.binlist = []
 
     def insert(self, value, priority=0):
-        ''' Add data element to the back of queue. '''
-        new_element = Element(value, priority)
-        current = self.front
+        ''' Add a value to the heap. '''
+        self.binlist.append(Element(value, priority))
+        self._promote(len(self.binlist)-1)
+        return
 
+    def _parent(self, index):
+        ''' Find the parents of a bin. '''
+        return [(index-1)//2, self.binlist[(index-1)//2].prio]
+
+    def _children(self, index):
+        ''' Find the children of a bin. Build list of children. '''
+        children = []
         try:
-            if new_element.prio > self.front.prio:
-                new_element.behind, self.front = self.front, new_element
-                return
-        except AttributeError:
-            self.front = new_element
-            return
+            child1 = (2 * index + 1, self.binlist[2*index+1].prio)
+        except IndexError:
+            pass
+        else:
+            children.append(child1)
 
-        current = self.front
-
-        while True:
-            try:
-                if new_element.prio <= current.behind.prio:
-                    current = current.behind
-                else:
-                    new_element.behind, current.behind = current.behind, new_element
-                    return
-            except AttributeError:
-                current.behind = new_element
-                return
+        try:
+            child2 = (2*index+2, self.binlist[2*index+2].prio)
```

Comment:
In this case, binding a local symbol to the index values you are using would allow you to write more concise code:

```python
idx1 = 2 * index + 1
idx2 = 2 * index + 2
children = []
for idx in (idx1, idx2):
    try:
        children.append((idx, self.binlist[idx].prio))
    except IndexError:
        continue
````